### ISSUE TITLE
Issue exporting text

### ISSUE THREAD

**Issue by Wolfbanus** [nico9889/OneNote2XournalppAddon#20]
Forgive me for sending you this issue on Christmas Eve but I've been wanting to forward this for a while but could never find the time.

- Blocks of text in onenote are exported truncated in xopp

![Schermata del 2024-12-24 16-12-24](https://github.com/user-attachments/assets/e15b5b2d-92dc-4635-8047-2112d96a2836)

![Schermata del 2024-12-24 16-14-06](https://github.com/user-attachments/assets/865d64e8-a087-4e2d-b74a-db0cfeb87d21)
&nbsp;
- Clicking on the text layer can both wrap and change the size of the text block and thereby solve the problem. 

![Schermata del 2024-12-24 16-19-02](https://github.com/user-attachments/assets/357fab5c-b8fe-46b9-a7af-ad7ab9ecce5d)

![Schermata del 2024-12-24 16-24-57](https://github.com/user-attachments/assets/559f84cd-ff0e-4d36-8729-a484cac3d471)
&nbsp;
- In the case of many blocks of text to be revised, however, the work becomes rather exhausting.

***
Buon Natale e buone feste!

**Comment by nico9889** [comment]
Hi, I attempted to reproduce the bug, but apparently I'm missing something
![immagine](https://github.com/user-attachments/assets/7980d52a-dcaf-439a-9f24-e2d0a28e32cb)

would you be able to share with me the exported XOPP?

If you don't want to share it publicly, please send it via email to cws@nico9889.me.

**Comment by nico9889** [comment]
I should've made more tests before answering, I wrote the text inserting manually all the "new line" trying to match your text format.
So I tried again rewriting it without splitting the sentences in different lines, and then I shrunk the text box.
In the end I obtained this:
![immagine](https://github.com/user-attachments/assets/b6fa7e6b-0e6c-40a3-acd5-960e88558322)

which is different from your problem, but I suspect the root cause is the same.

I'm looking into this and I'll let you know when I'll have a new version ready to test, so you can try it and see if your problem will be solved as well.

**Comment by nico9889** [comment]
It looks like it's going to take some time: OneNote is splitting the paragraphs in """weird""" chunks like this one
![immagine](https://github.com/user-attachments/assets/c39eb3ac-a28b-4c06-9ee3-a527ccf18a3d)

All I know about that chunk is it's content (the smaller, highlighted rectangle) **and the paragraph size** (the wider dashed rectangle).

---

TL;DR: Xournal++ has limited text capabilities. Try for example to create a single text with multiple colors: you can't, you need to split it into multiple texts. I'm facing a similar issues with paragraphs, which are already split by OneNote in a weird way.

---

The problem is that typography really is hard (https://medium.engineering/typography-is-impossible-5872b0c7f891).
OneNote is delegating the problem to the browser renderer, which is taking into account **both the text chunk size and the paragraph size to split the text**.
I'm currently cheating exporting the texts using a "monospace" font, which it has a fixed character size, so I can calculate easily the text length and split it by myself, roughly emulating what the browser does.

This approach works as long as the text is in a single chunk that starts with the paragraph, but a chunk in the middle that wraps itself it's horribly hard to handle, because:
1) I lack some information (where does the text starts in coordinates?);
2) Xournal++ text editor is poor and doesn't support nor paragraph neither text chunks.

The first problem is "easy" to solve, I just need to calculate the length of the previous chunk, so this way I get where the next chunk starts. 

The second problem is harder, and as of now only two solutions came up to my mind:
* creating a text box of the size of the paragraph and filling it with empty spaces to match the text chunk position (which is a **horrible** solution in my opinion);
* splitting the wrapped text chunks into multiple texts boxes (which is a more correct approach, but harder to implement and worse to edit once exported, but it's a Xournal++ limitation).

---

Volevo ricambiare gli auguri una volta risolto il problema, ma di questo passo finirei per farli l'anno prossimo nel caso peggiore, pertanto buon Natale e buone feste anche a te.

**Comment by nico9889** [comment]
I have a both a good news and a bad news.

The good news is that I was working behind the wrong assumption that the browser only let me know the width of the rectangle that included the wrapped text (the dashed rectangle in the screenshot in the previous message). I discovered that I can access the separate position of the text on the first line and the second line, which simplified the work a lot.

The bad news is that even if I know the position of the text, I still need to guess which text goes on the first line and which on the second line.

I'll make an example to explain myself better:
![immagine](https://github.com/user-attachments/assets/8df0ba9d-1442-4a24-84e7-7e09481c9cde)
Here I have the text "isurare, che mi interessa misurare." and I need to calculate how many characters still fit the upper rectangle with the highlighted background and put the remaining on the second line.

Sometimes it works:
![immagine](https://github.com/user-attachments/assets/ae704134-0860-4f36-a694-9c0dfb3fd3b0)

Sometimes it doesn't but it's "acceptable" (I'm trying not to split words, the word "hermitiano" is splitted by OneNote):
![immagine](https://github.com/user-attachments/assets/2dc7d57c-bf60-4893-9b73-11e97663c45d)

Sometimes it's overlapped and unreadable:
![immagine](https://github.com/user-attachments/assets/ccf3c46e-94dd-4759-9e15-a9c3da2b8b6c)

I've published a pre-release here on GitHub, you have the instructions on how to install it on Firefox here:
https://github.com/nico9889/OneNote2XournalppAddon/issues/17#issuecomment-2424233343

Try it when you have time and let me know if it solved at least partially the issue you were reporting.

**Comment by nico9889** [comment]
Hello again,
I updated the pre-release file 😅 with a very hacky workaround I almost reached the perfection:
![immagine](https://github.com/user-attachments/assets/f931505b-7f24-46e7-83cf-ce8d6ea6802f)

To find where the text goes to a new line I need to change **temporarily** the content of the page.
This **may** break the OneNote web app functionality, until refreshing the page.
**Please let me know if you encounter some issues, this time is very important.**

Unfortunately this still breaks when changing the font from the default one.
I'm planning to map the OneNote most frequently used fonts to a compatible fonts in Xournal++, but this is a very time expensive operation and I won't do it soon.

**Comment by Wolfbanus** [comment]
I've tried the 1.8.1 version on the same onenote file

I will send you the .one and .xopp files via email because github doesn't allow me to upload here (maybe there's a way but i don't know it)

In summary, text export is much better but there are still few bugs.

***
Grazie per gli auguri e per l'impegno che ci metti nel sviluppare questo addon!

**Comment by nico9889** [comment]
> I will send you the .one and .xopp files via email...

I answered your email, please check your spam.

> ... because github doesn't allow me to upload here (maybe there's a way but i don't know it)

It seems like GitHub supports only a limited set of file including the ZIP format, if you need it the next time try to zip the file if unsupported.

**Comment by nico9889** [comment]
Hi, in the end I found a way to export the text without touching the page content, so I anticipated the release.

The issue is that I discovered a bigger problem with the fonts that apparently I have no way to solve (this time for real), unless Firefox and Chrome change themselves.

If you are interested I attempted to explain it here: https://github.com/nico9889/OneNote2XournalppAddon/wiki/The-Font-Dilemma.

In short: there's a font compatibility issue, if you see anything strange in the output file then try to change the font or try to export the file from a computer with another operating system.

I'll close this issue for now as it's "solved". I may open a more specific issue about the font in the future.
